controllers/resources/lease.go

```diff
@@ -0,0 +1,272 @@
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.calculateLeaseDurationBySingleRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
+		m.log.Error(err, "couldn't obtain node lease node error getting node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		l, _ := m.commonLeaseManager.GetLease(context.Background(), node)
+		if l != nil && l.Spec.HolderIdentity != nil && *l.Spec.HolderIdentity != holderIdentity {
+			m.log.Info("can't acquire node lease, it is already owned by another owner", "current owner", *l.Spec.HolderIdentity)
+			return false, &RequeueIfLeaseTaken, nil
+		}
+
+		m.log.Error(err, "couldn't obtain lease for node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	//all good lease created with wanted duration
+	return true, &leaseDuration, nil
+
+}
+
+func (m *nhcLeaseManager) calculateLeaseDurationBySingleRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
+	leaseDuration := DefaultLeaseDuration
+	if len(nhc.Spec.EscalatingRemediations) == 0 {
+		return DefaultLeaseDuration
+	} else {
+		remediationKind := remediationCR.GetKind()
+		for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+			if strings.TrimSuffix(esRemediation.RemediationTemplate.Kind, "Template") == remediationKind {
+				leaseDuration = esRemediation.Timeout.Duration
+				break
+			}
+		}
+	}
+	return leaseDuration
+}
+
+func (m *nhcLeaseManager) calculateLeaseDurationByAllRemediations(ctx context.Context, node *v1.Node, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error) {
+	highestRemediationLeaseDuration := DefaultLeaseDuration
+	if len(nhc.Spec.EscalatingRemediations) > 0 {
+
+		remediationCrs, err := listRemediationCRs(ctx, m.client, nhc, func(cr unstructured.Unstructured) bool {
+			return cr.GetName() == node.GetName()
+		})
+		if err != nil {
+			m.log.Error(err, "couldn't fetch remediation CRs in order to calculate required lease duration", "node name", node.Name)
+			return 0, err
+		}
+
+		highestRemediationLeaseDuration = 0
+		for _, remediationCr := range remediationCrs {
+			if currentLeaseDuration := m.calculateLeaseDurationBySingleRemediation(&remediationCr, nhc); currentLeaseDuration > highestRemediationLeaseDuration {
+				highestRemediationLeaseDuration = currentLeaseDuration
+			}
+		}
+
+	}
+	return highestRemediationLeaseDuration, nil
+}
+
+// needs to release leases that needs to be released and extend those that needs extending return minimum timeout for next check
+func (m *nhcLeaseManager) manageLeases(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error) {
```

Comment:
I didn't understand yet what this thing is doing, but iterating all nodes should be avoided IMHO. Will try give better feedback with idea what to change in the next round of review...

Reply:
In a nutshell, the nodes are used to find all potential leases that may exist, than each found lease is managed separately (i.e extended, deleted and getting the closest requeue)